lib/Doctrine/ORM/Cache/DefaultQueryCache.php

```diff
@@ -210,6 +210,25 @@ public function get(QueryCacheKey $key, ResultSetMapping $rsm, array $hints = []
                 $collection->setInitialized(true);
             }
 
+            foreach ($data as $fieldName => $unCachedAssociationData) {
+                // In some scenarios, such as EAGER+ASSOCIATION+ID+CACHE, the
+                // cache key information in `$cacheEntry` will not contain details
+                // for fields that are associations.
+                //
+                // This means that `$data` keys for some associations that may
+                // actually not be cached will not be converted to actual association
+                // data, yet they contain L2 cache AssociationCacheEntry objects.
+                //
+                // We need to unwrap those associations into proxy references,
+                // since we don't have actual data for them except for identifiers.
+                if ($unCachedAssociationData instanceof AssociationCacheEntry) {
```

Comment:
It makes sense IMO, but I feel like we could do that conversion while iterating `$entry['associations']` in https://github.com/doctrine/doctrine2/pull/6640/files#diff-c5343cb0b202102a59de56dfdcb8ee47R147 to avoid iterating the entire object data.

Reply:
`$entry['associations']` does not contain the keys we're iterating upon here.

I considered using an accumulator of already processed keys, but since these are usually about associations, and the number of fields and associations is (typically) < 10 on any entity, looping is actually faster than aggregating the keys and excluding them here.